### ISSUE TITLE
ProjectReferences disappear if you first build GetCopyToOutputDirectoryItems

### ISSUE THREAD

**Issue by KirillOsenkov** [dotnet/msbuild#4923]
```
md horrifying
cd horrifying
dotnet new classlib -o A
dotnet new classlib -o B
dotnet add B reference A
dotnet new sln
dotnet sln add A
dotnet sln add B
cd B
msbuild /bl /t:GetCopyToOutputDirectoryItems;Build
```

When B is being compiled the project reference to A is dropped, so csc.exe doesn't get a reference to A.dll if you first build `GetCopyToOutputDirectoryItems` target before building the `Build` target.

This is because `_SplitProjectReferencesByFileExistence` logically depends on `AssignProjectConfiguration` which creates the `@(ProjectReferenceWithConfiguration)` item that _Split uses as input:

```
    <PrepareProjectReferencesDependsOn>
      AssignProjectConfiguration;
      _SplitProjectReferencesByFileExistence;
      _GetProjectReferenceTargetFrameworkProperties
    </PrepareProjectReferencesDependsOn>
```

However `GetCopyToOutputDirectoryItemsDependsOn` doesn't declare a dependency on `AssignProjectConfiguration` for _Split, and so at the time _Split first runs the `@(ProjectReferenceWithConfiguration)` item is empty:

```
    <GetCopyToOutputDirectoryItemsDependsOn>
      AssignTargetPaths;
      _SplitProjectReferencesByFileExistence;
      _GetProjectReferenceTargetFrameworkProperties
    </GetCopyToOutputDirectoryItemsDependsOn>
```

Next during the actual Build the target `_SplitProjectReferencesByFileExistence` is skipped because it was previously built successfully, and so the project references never get added.

See also https://github.com/KirillOsenkov/MSBuildStructuredLog/issues/269

**Comment by KirillOsenkov** [comment]
https://github.com/microsoft/msbuild/issues/943 might or might not be potentially related

**Comment by KirillOsenkov** [comment]
FYI @cston I've been investigating your issue with a project reference to System.Windows.Forms disappearing and this looks like the underlying root cause. In your log the `GetCopyToOutputDirectoryItems` target gets built first; at which time `AssignProjectConfiguration` hasn't run yet, so `@(ProjectReferenceWithConfiguration)` is empty, and it's memoized so next time it wants to run it just skips, never adding the project references.

**Comment by KirillOsenkov** [comment]
It's also bad that `GetCopyToOutputDirectoryItemsDependsOn` stomps over previous property value without including itself, it makes it difficult to change its value.

I had to do this for a workaround:
```
<Project>

  <Import Project="Sdk.Props" Sdk="Microsoft.NET.Sdk" />

  <ItemGroup>
    <ProjectReference Include="..\A\A.csproj" />
  </ItemGroup>

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>

  <Import Project="Sdk.Targets" Sdk="Microsoft.NET.Sdk" />

  <PropertyGroup>
    <GetCopyToOutputDirectoryItemsDependsOn>AssignProjectConfiguration;$(GetCopyToOutputDirectoryItemsDependsOn)</GetCopyToOutputDirectoryItemsDependsOn>
  </PropertyGroup>

</Project>
```

**Comment by KirillOsenkov** [comment]
Of course there's this: https://github.com/microsoft/msbuild/issues/1468

**Comment by KirillOsenkov** [comment]
@cdmihai it looks like you may have already fixed this recently:
https://github.com/microsoft/msbuild/pull/4865

Could you verify that this bug would be fixed with your change?

**Comment by KirillOsenkov** [comment]
Is this related to https://github.com/microsoft/msbuild/issues/4677?

**Comment by rainersigwald** [comment]
After https://github.com/dotnet/project-system/pull/5898, this is causing design-time build failures in the Roslyn repo at least (AB#1073852). So it wasn't fixed by #4865.

**Comment by cdmihai** [comment]
#4865 does not change behaviour by default, because who knows what blows up. You need to set `$(MSBuildCopyContentTransitively )` to `true` to force transitive item copying, even in VS.

**Comment by KirillOsenkov** [comment]
Hold on I thought we need to add 
`<GetCopyToOutputDirectoryItemsDependsOn>AssignProjectConfiguration;... ` regardless? I thought it's an underspecified dependency that still needs to be made explicit?

**Comment by cdmihai** [comment]
#4865 adds that dependency when you set `$(MSBuildCopyContentTransitively)` to true. Otherwise it fallbacks to legacy behaviour and does nothing.

**Comment by KirillOsenkov** [comment]
I think we should always set this?

**Comment by cdmihai** [comment]
The original idea was to always force recursive content copying in the next major VS version, in case it's a bigger breaking than expected, and until then, it's on demand by whoever is affected.

**Comment by KirillOsenkov** [comment]
I see.

I'm guessing we want to do this orthogonally, fix the problem now and turn on recursive copying later. Too many people are hitting this.

My recommendation!

**Comment by cdmihai** [comment]
> I see.
> 
> I'm guessing we want to do this orthogonally, fix the problem now and turn on recursive copying later. Too many people are hitting this.
> 
> My recommendation!

By this you mean fix content discovery to always be transitive (by adding the undeclared target dependencies), but keep content copying with the same legacy behaviour and a property to force the various types of content copying? Might be tricky to separate them ...

**Comment by KirillOsenkov** [comment]
I got confused, I thought transitive content copying is separate from declaring an explicit dependency on AssignProjectConfiguration. But it looks like the actual fact of declaring that GetCopyToOutputDirectoryItemsDependsOn depends on AssignProjectConfiguration turns on transitive copying? Is my understanding correct?

In that case I'd wager let's just do it and turn it on by default. It doesn't matter when to break people, I'd say the sooner the better. The non-deterministic inexplicable behavior of some items being copied sometimes but not on others is terrible. Most of the time from my experience it's already transitive (due to the fact that AssignProjectConfiguration usually runs before GetCopyToOutputDirectoryItems). When that's not the case you're pretty likely to get weird errors like this issue.

In general, the logic of which items get copied to output and which not from my experience has caused a tremendous amount of pain and friction in our ecosystem. The more deterministic, predictable and documented we make it, the better. 

A lot of people are optimizing their builds right now, and we don't want them to optimize against buggy defaults that will change anyway. Let's break it soon, let's document the heck out of this, and let's strive to make content copying as predictable and deterministic as possible. Also easy to control. And I want a pony too.

**Comment by KirillOsenkov** [comment]
I literally thought we should do this:

![image](https://user-images.githubusercontent.com/679326/75508622-9a311e00-5999-11ea-9883-a3323831aae0.png)

**Comment by cdmihai** [comment]
> I got confused, I thought transitive content copying is separate from declaring an explicit dependency on AssignProjectConfiguration. But it looks like the actual fact of declaring that GetCopyToOutputDirectoryItemsDependsOn depends on AssignProjectConfiguration turns on transitive copying? Is my understanding correct?

That's correct, adding the dependency makes content copying transitive. @rainersigwald for the decision to force transitive copying on now, rather than next VS release.

**Comment by rainersigwald** [comment]
Team triage: we're all interested in just fixing the problem (setting `MSBuildCopyContentTransitively=true`). We're concerned about VS perf, which we expect may be particularly sensitive to global-property mismatch.

@davkean, can you have someone look at the project-system perf impact of setting that property?

We also need to decide where to set the property. If we set it in Common.targets, it'll apply to all project types, while the fix is only necessary in CPS after the FUTD fix. So maybe we should set this only in the .NET Core SDK.

**Comment by ezverev** [comment]
> It's also bad that `GetCopyToOutputDirectoryItemsDependsOn` stomps over previous property value without including itself, it makes it difficult to change its value.
> 
> I had to do this for a workaround:
> 
> ```
> <Project>
> 
>   <Import Project="Sdk.Props" Sdk="Microsoft.NET.Sdk" />
> 
>   <ItemGroup>
>     <ProjectReference Include="..\A\A.csproj" />
>   </ItemGroup>
> 
>   <PropertyGroup>
>     <TargetFramework>netstandard2.0</TargetFramework>
>   </PropertyGroup>
> 
>   <Import Project="Sdk.Targets" Sdk="Microsoft.NET.Sdk" />
> 
>   <PropertyGroup>
>     <GetCopyToOutputDirectoryItemsDependsOn>AssignProjectConfiguration;$(GetCopyToOutputDirectoryItemsDependsOn)</GetCopyToOutputDirectoryItemsDependsOn>
>   </PropertyGroup>
> 
> </Project>
> ```

Is there a possible workaround for this for VS 2017 15.9.20? (MSBuild 15.0)
I tried to apply the described workaround and failed.

**Comment by BenVillalobos** [comment]
Turns out this is fixed in 5.0+ SDK's via https://github.com/dotnet/sdk/pull/12318, and fixed in 17.0+ MSBuild's via https://github.com/dotnet/msbuild/pull/6622.

**Comment by KirillOsenkov** [comment]
Is #943 fixed as well?

**Comment by KirillOsenkov** [comment]
And #1845?

**Comment by BenVillalobos** [comment]
> Is #943 fixed as well?

This should be fixed, added to the 17.0 milestone.

> And #1845?

Stil repros. not sure of the details though.

**Comment by drewnoakes** [comment]
We are looking to reinstate a version of the .NET Project System change that originally flagged this issue.

https://github.com/dotnet/project-system/pull/7932

I will test the change against Roslyn to make sure the design time builds are not impacted.

Any concerns?
